Add coverage tools

Martolivna opened the pull request:
This PR adds a coverage tools, which assist in monitoring test coverage.

- [x] Update requirements
- [x] Update the CI
- [ ] Integrate [Codecov](https://codecov.io/)

webknjaz commented on mysite/settings.py:
```diff
@@ -32,6 +32,7 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'coverage',
```
What is this app? Where does it come from?

Martolivna replied:
I accidentally forgot about the link you shared for coverage.py and ended up following this one: https://realpython.com/testing-in-django-part-1-best-practices-and-examples/

webknjaz replied:
That's weird. It says to add it but doesn't show it as a code example. I wasn't able to find any examples of doing so on the internet. Does it work? Does anything change if you remove it?

Martolivna replied:
I've checked it, and it seems like it really doesn't affect anything. I've made changes https://github.com/kpi-web-guild/django-girls-blog-Martolivna/pull/21/commits/f5f6ceeeeae84a95111d275b0669193609030367 .

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/dev.in:
```diff
@@ -1,2 +1,3 @@
 -r base.txt
 pylint
+coverage
```
I recommend to keep the entries sorted for maintenability.

Martolivna replied:
Addressed https://github.com/kpi-web-guild/django-girls-blog-Martolivna/pull/21/commits/f5f6ceeeeae84a95111d275b0669193609030367 .

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/dev.txt:
```diff
@@ -4,37 +4,39 @@
 #    pip-compile --constraint=base.in dev.in
 #
-asgiref == 3.7.2
+asgiref==3.7.2
```
It looks like you used a newer version of pip-tools this time and it produced extra formatting changes. I recommend submitting them separately so that the new dependency addition is easy to spot in the diff. Typically, it is forbidden to mix functional and formatting changes.

Martolivna replied:
Earlier, it was also generated without spaces. You recommended adding spaces for improved readability, and I manually included them. Given that it remains a generated file, I refrained from adding spaces.

webknjaz replied:
So I wasn't clear, then. Spaces do indeed improve readability and I recommend adding them to input files that you are in control of. But as soon as we delegated the constraints file management to pip-tools, they stopped being human-maintained.
Whenever things are regenerated by a tool with pre-defined output formatting, we should stop applying any other changes since they'll be overrided by said tool.

To summarize, use this advice in `.in` files but not `.txt`, when using pip-tools. When it's not used and you manage the file by hand, you can make the formatting nicer regardless.

Martolivna replied:
That's exactly what I think, so I stopped editing. Thank you!

webknjaz replied:
Marking this as resolved.

webknjaz replied:
Let's move the formatting to a separate PR, it's confusing that most of the patch is formatting changes with any other changes being hard to locate.

webknjaz reviewed (commented):
When integrating test-related things, it's best to make sure they are actually executed.
Could you update the GHA job to generate coverage when it invokes tests?
Also, could you integrate https://codecov.io so that the coverage is uploaded to that service from CI.

Martolivna committed:
Add a coverage tool

It assists in monitoring test coverage.
Ref #21

Martolivna committed:
Update requirements formatting

Ref #21

Martolivna committed:
Update the tests GHA job to generate coverage

Ref #21

Martolivna committed:
Integrate Codecov via GHA

This commit updates the CI job for tests to upload a coverage report
to Codecov.

Ref #21

Martolivna commented:
@webknjaz Could you please add a Codecov secret token as repository secret to action secret and variables in this repository? You can get it [there](https://app.codecov.io/gh/kpi-web-guild/django-girls-blog-Martolivna/new).

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -9,37 +9,42 @@ jobs:
-      uses: actions/setup-python@v4
-      with:
-        python-version: 3.11
+      - name: Fetch the src
```
Let's not make formatting changes in the same place where functional changes are made. It generates a hard to review diff every time this happens. Never do so. Also, if you want to lint YAML, make use of https://github.com/cherrypy/cheroot/blob/main/.yamllint config. This of course, must be done separately from any functional changes.

Martolivna replied:
This formatting changes was made automatically. I didn't notice that.

webknjaz replied:
You need to understand your development environment configuration and be in control of it. Check your editor and separate the changes.
Also, check what you commit. Using `git commit -v` will show you the preview of what is going into the patch. Check the PR's files tab too — there's no need for an external reviewer to be checking this for you since you can spot it faster and save yourself some time by reducing the number of review rounds.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -9,37 +9,42 @@ jobs:
+      - name: Upload coverage reports to Codecov
+        uses: codecov/codecov-action@v3
+        env:
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```
This token is unnecessary for public repos.
```suggestion
```

webknjaz replied:
I guess you can disregard this per https://github.com/kpi-web-guild/django-girls-blog-Martolivna/pull/21#issuecomment-1723553104.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -9,37 +9,42 @@ jobs:
+      - name: Run pre-commit
+        run: python -m pre_commit run --all-files
+        env:
+          SKIP: no-commit-to-branch
```
Don't do this, per the above explanation.

webknjaz commented on requirements/base.txt:
```diff
@@ -4,11 +4,11 @@
 #    pip-compile --constraint=dev.txt base.in
 #
-asgiref == 3.7.2
+asgiref==3.7.2
```
Same here. The PR title says "add a coverage app".

webknjaz commented:
> @webknjaz Could you please add a Codecov secret token as repository secret to action secret and variables in this repository? You can get it [there](https://app.codecov.io/gh/kpi-web-guild/django-girls-blog-Martolivna/new).

This is not technically necessary but having seen instability problems in the past, I've added the secret.

webknjaz commented:
@Martolivna you're no longer adding a Django app in this PR so it probably needs a better title and description.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -9,37 +9,42 @@ jobs:
-    - name: Run tests
-      run: python -m manage test
+      - name: Run tests
+        run: python -m coverage run manage.py test -v 2
```
I like how https://adamj.eu/tech/2019/04/30/getting-a-django-application-to-100-percent-coverage/ integrates the coverage initialization into `manage.py`. This allows running the same simple command in CI and locally without having to duplicate the coverage wrapper invocation.

Although, I'd probably structure that as a context manager, making use of `@contextlib.contextmanager` and `contextlib.nullcontext()`. It'd add more elegance to the integration.

webknjaz replied:
Per https://coverage.readthedocs.io/en/latest/api_coverage.html, calling `Coverage.start()` and `Coverage.stop()` separately can be replaced by the `Coverage.collect()` context manager. So I'd use that too.

webknjaz commented:
@Martolivna any progress on this?

webknjaz commented:
@Martolivna time to resume this?

Martolivna commented:
> @Martolivna time to resume this?

@webknjaz Exactly! Thank you.
